simulator/simulator.go

```diff
@@ -0,0 +1,197 @@
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/bitrise-io/go-utils/command"
+	"github.com/bitrise-io/go-utils/env"
+	"github.com/bitrise-io/go-utils/errorutil"
+	"github.com/bitrise-io/go-utils/log"
+	sim "github.com/bitrise-io/go-xcode/simulator"
+)
+
+// Info ...
+type Info sim.InfoModel
+
+// Manager ...
+type Manager interface {
+	GetLatestSimulatorInfoAndVersion(osName, deviceName string) (Info, string, error)
+	GetSimulatorInfo(osNameAndVersion, deviceName string) (Info, error)
+	LaunchSimulator(simulatorID string, xcodebuildMajorVersion int) error
+
+	ResetLaunchServices() error
+	SimulatorBoot(id string) error
+	SimulatorEnableVerboseLog(id string) error
+	SimulatorCollectDiagnostics() (string, error)
+	SimulatorShutdown(id string) error
+	SimulatorDiagnosticsName() (string, error)
+}
+
+type manager struct {
+}
+
+// NewManager ...
+func NewManager() Manager {
+	return manager{}
+}
+
+func (s manager) GetLatestSimulatorInfoAndVersion(osName, deviceName string) (Info, string, error) {
+	info, ver, err := sim.GetLatestSimulatorInfoAndVersion(osName, deviceName)
+	return Info(info), ver, err
+}
+
+func (s manager) GetSimulatorInfo(osNameAndVersion, deviceName string) (Info, error) {
+	info, err := sim.GetSimulatorInfo(osNameAndVersion, deviceName)
+	return Info(info), err
+}
+
+func (s manager) LaunchSimulator(simulatorID string, xcodebuildMajorVersion int) error {
+	return sim.BootSimulator(simulatorID, xcodebuildMajorVersion)
+}
+
+// Reset launch services database to avoid Big Sur's sporadic failure to find the Simulator App
+// The following error is printed when this happens: "kLSNoExecutableErr: The executable is missing"
+// Details:
+// - https://stackoverflow.com/questions/2182040/the-application-cannot-be-opened-because-its-executable-is-missing/16546673#16546673
+// - https://ss64.com/osx/lsregister.html
+func (s manager) ResetLaunchServices() error {
+	f := command.NewFactory(env.NewRepository())
+	cmd := f.Create("sw_vers", []string{"-productVersion"}, nil)
+
+	macOSVersion, err := cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return err
+	}
+
+	if strings.HasPrefix(macOSVersion, "11.") { // It's Big Sur
+		cmd := f.Create("xcode-select", []string{"--print-path"}, nil)
+		xcodeDevDirPath, err := cmd.RunAndReturnTrimmedCombinedOutput()
+		if err != nil {
+			return err
+		}
+
+		simulatorAppPath := filepath.Join(xcodeDevDirPath, "Applications", "Simulator.app")
+
+		cmdString := "/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Support/lsregister"
+		cmd = f.Create(cmdString, []string{"-f", simulatorAppPath}, nil)
+
+		log.Infof("Applying launch services reset workaround before booting simulator")
+		_, err = cmd.RunAndReturnTrimmedCombinedOutput()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s manager) SimulatorBoot(id string) error {
+	f := command.NewFactory(env.NewRepository())
```

Comment:
`Factory` should be injected here.

Reply:
I only broke the `step` and `xcodebuild` package's dependencies fully, other packages (including simulator) are not fully migrated yet.